mobius-core/src/main/java/com/spotify/mobius/MobiusLoop.java

```diff
@@ -184,12 +185,25 @@ public void dispose() {
 
   @Override
   public void dispose() {
-    synchronized (modelObservers) {
+    synchronized (disposeLock) {
+      synchronized (modelObservers) {
+        // Remove model observers so that they receive no further model changes.
+        modelObservers.clear();
+      }
+
+      // Disable the event and effect dispatchers. This will cause any further
+      // events or effects posted to the dispatchers to silently be ignored.
+      eventDispatcher.disable();
+      effectDispatcher.disable();
+
+      // Stop the event source and effect handler.
+      eventSourceDisposable.dispose();
```

Comment:
Why not do this first of all? That would slightly reduce the risk of lost events and would open up for having a 'finishing event' in an `EventSource`. Which feels a bit like an antipattern, but might be something someone would have a use for at some point. My suggested order would then be: 

1. dispose event source
1. disable event processing
1. disable effects
1. clear observers (can go before disabling effects)
1. dispose effect consumer
1. dispose event dispatcher
1. dispose effect dispatcher
1. mark as disposed (this could maybe be moved up to right after disabling event processing, but that's not likely to have any impact in practice)